src/instance.rs

```diff
@@ -137,137 +136,95 @@ pub fn prepare_module(
+                "module does not have a wasi start function".to_string(),
+            ))?;
 
-                    debug!("notifying main thread we are about to start");
-                    tx.send(Ok(())).unwrap();
+        debug!("starting wasi instance");
 
-                    debug!("starting wasi instance");
+        let cg = oci::get_cgroup(&spec)?;
 
-                    // TODO: How to get exit code?
-                    // This was relatively straight forward in go, but wasi and wasmtime are totally separate things in rust.
-                    let (lock, cvar) = &*exit_code;
-                    let _ret = match f.call(&mut store, &mut [], &mut []) {
-                        Ok(_) => {
-                            debug!("exit code: {}", 0);
-                            let mut ec = lock.lock().unwrap();
-                            *ec = Some((0, Utc::now()));
-                        }
-                        Err(_) => {
-                            error!("exit code: {}", 137);
-                            let mut ec = lock.lock().unwrap();
-                            *ec = Some((137, Utc::now()));
-                        }
-                    };
+        oci::setup_cgroup(&cg, &spec)?;
 
-                    cvar.notify_all();
-                })?;
+        let res = unsafe { exec::fork(Some(&cg)) }?;
+        match res {
+            exec::Context::Parent(tid, pidfd) => {
+                let mut lr = self.pidfd.lock().unwrap();
+                *lr = Some(pidfd.clone());
+
+                debug!("started wasi instance with tid {}", tid);
 
-        debug!("Waiting for start notification");
-        match rx.recv().unwrap() {
-            Ok(_) => (),
-            Err(err) => {
-                debug!("error starting instance: {}", err);
                 let code = self.exit_code.clone();
 
-                let (lock, cvar) = &*code;
-                let mut ec = lock.lock().unwrap();
-                *ec = Some((139, Utc::now()));
-                cvar.notify_all();
-                return Err(err);
+                let _ = thread::spawn(move || {
+                    let (lock, cvar) = &*code;
+                    let status = match pidfd.wait() {
+                        Ok(status) => status,
+                        Err(e) => {
+                            error!("error waiting for pid {}: {}", tid, e);
+                            cvar.notify_all();
+                            return;
+                        }
+                    };
+
+                    debug!("wasi instance exited with status {}", status.status);
+                    let mut ec = lock.lock().unwrap();
+                    *ec = Some((status.status, Utc::now()));
+                    drop(lock);
+                    cvar.notify_all();
+                });
+                Ok(tid)
+            }
+            exec::Context::Child => {
+                // child process
+
+                // TODO: How to get exit code?
```

Comment:
I believe it's totally fine to assume that Okay case returns exit code 0. The trick case is how to handle error. 

See [here](https://docs.wasmtime.dev/api/wasmtime/struct.Func.html#errors-4), wasmtime `call()` will return an `Error` type which can be `Trap` or type error in params. 

[Trap](https://docs.wasmtime.dev/api/wasmtime/enum.Trap.html) is defined as 

```rust
pub enum Trap {
    StackOverflow,
    MemoryOutOfBounds,
    HeapMisaligned,
    TableOutOfBounds,
    IndirectCallToNull,
    BadSignature,
    IntegerOverflow,
    IntegerDivisionByZero,
    BadConversionToInteger,
    UnreachableCodeReached,
    Interrupt,
    AlwaysTrapAdapter,
    OutOfFuel,
}
```

Maybe we can try mapping each enum value of `Trap` to proper exit code?

Reply:
From what I recall when working with wasmtime in go, was seems to support exit codes.
We could try to capture these traps and translate though.

This isn't new code though, just moved due to whitespace changes.